README.md
# Itamae::Plugin::Recipe::Rbenv

[Itamae](https://github.com/ryotarai/itamae) plugin to install ruby with rbenv

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'itamae-plugin-recipe-rbenv'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install itamae-plugin-recipe-rbenv

# Usage
## System wide installation

Install rbenv to /usr/local/rbenv or some shared path

### Recipe

```ruby
# your recipe
include_recipe "rbenv::system"
```

### Node

Use this with `itamae -y node.yml`

```yaml
# node.yml
rbenv:
  global:
    2.3.0
  versions:
    - 2.3.0
    - 2.2.4

  # rbenv install dir, optional (default: /usr/local/rbenv)
  rbenv_root: "/path/to/rbenv"

  # specify scheme to use in git clone, optional (default: https)
  scheme: git

  # Create /usr/local/rbenv/cache, optional (default: false)
  # See: https://github.com/rbenv/ruby-build#package-download-caching
  cache: true

  # Whether install dependencies (default: true)
  # Recommend false if `--no-sudo`
  install_dependency: true

  # Install arbitrary rbenv plugins, optional (default: [])
  plugins:
    dcarley/rbenv-sudo:
      revision: master

# ruby-build is always installed. Specifying revision improves performance.
ruby-build:
  revision: e455975286e44393b1b33037ae1ce40ef2742401

# Optional plugin. Specify :install or :revision to install rbenv-default-gems.
rbenv-default-gems:
  default-gems:
    - bundler
    - bcat ~>0.6
    - rails --pre
  install: true
  # or
  revision: ead67889c91c53ad967f85f5a89d986fdb98f6fb
```

### .bashrc

Recommend to append this to .bashrc in your server.

```bash
export RBENV_ROOT=/usr/local/rbenv
export PATH="${RBENV_ROOT}/bin:${PATH}"
eval "$(rbenv init -)"
```

## Installation for a user

Install rbenv to `~#{node[:rbenv][:user]}/.rbenv`

### Recipe

```ruby
# your recipe
include_recipe "rbenv::user"
```

### Node

Use this with `itamae -y node.yml`

```yaml
# node.yml
rbenv:
  user: k0kubun
  global:
    2.3.0
  versions:
    - 2.3.0
    - 2.2.4

  # specify scheme to use in git clone, optional (default: https)
  scheme: git

  # Create ~/.rbenv/cache, optional (default: false)
  # See: https://github.com/rbenv/ruby-build#package-download-caching
  cache: true

  # Install build dependencies or not (default: true)
  # Recommend false if `--no-sudo`
  install_dependency: true

  # Install dependencies to build *-dev or not (default: false)
  install_development_dependency: false

# ruby-build is always installed. Specifying revision improves performance.
ruby-build:
  revision: e455975286e44393b1b33037ae1ce40ef2742401

# Optional plugin. Specify :install or :revision to install rbenv-default-gems.
rbenv-default-gems:
  default-gems:
    - bundler
    - bcat ~>0.6
    - rails --pre
  install: true
  # or
  revision: ead67889c91c53ad967f85f5a89d986fdb98f6fb
```

## MItamae

This plugin can be used for MItamae too. Put this repository under `./plugins` as git submodule.

```rb
node.reverse_merge!(
  rbenv: {
    user: 'k0kubun',
    global: '2.3.1',
    versions: %w[
      2.3.1
      2.2.5
    ],
  }
)

include_recipe "rbenv::user"
```

## License

MIT License


CHANGELOG.md
## v0.9.5
### Added
- Supported Raspbian

## v0.9.4
### Fixed
- Fixed CentOS 7- to install libyaml-devel

## v0.9.3
### Fixed
- Fixed a bug in CentOS's version check

## v0.9.2
### Fixed
- Fixed the condition to install rust on CentOS

## v0.9.1
### Added
- Added gmp and rustc as dependencies

## v0.9.0
### Changed
- [breaking change] Default `scheme` was changed from git to https

## v0.8.6
### Fixed
- Skip installing libyaml-devel on CentOS / RedHat 8+

## v0.8.5
### Added
- Install missing dependencies for openSUSE leap 15

## v0.8.4
### Added
- Install missing dependencies on openSUSE leap

## v0.8.3
### Added
- support FreeBSD and OpenBSD

## v0.8.2
### Added
- Support Ubuntu 20.04
- Support Gentoo

## v0.8.1
### Added
- Support Debian Testing

## v0.8.0
### Changed
- [breaking change] Optionalize development dependencies. Set `install_development_dependency: true` to enable them.

## v0.7.3
### Added
- Support Debian 10 buster
- Install subversion for source build dependency

## v0.7.2
### Added
- Install `autoconf` and `bison` on RedHat platforms [#26](https://github.com/k0kubun/itamae-plugin-recipe-rbenv/pull/26)
  *Thanks to @hsbt*

## v0.7.1
### Added
- Install `make` on RedHat platforms [#25](https://github.com/k0kubun/itamae-plugin-recipe-rbenv/pull/25)
  *Thanks to @hsbt*

## v0.7.0
### Changed
- [breaking change] Raise NotImplementedError on an unknown platform
- Initial support of OpenBSD

## v0.6.11
### Fixed
- Avoid unnecessary `rbenv global` [#24](https://github.com/k0kubun/itamae-plugin-recipe-rbenv/pull/24)
  *Thanks to @pocke*

## v0.6.10
### Fixed
- Fix undefined variable error and installation requirement in `node[:rbenv][:plugins]` feature [#23](https://github.com/k0kubun/itamae-plugin-recipe-rbenv/pull/23)
  *Thanks to @mozamimy*

## v0.6.9
### Added
- Add `node[:rbenv][:plugins]` to install arbitrary rbenv plugins [#21](https://github.com/k0kubun/itamae-plugin-recipe-rbenv/pull/21)
  *Thanks to @Yuki-Inoue*

## v0.6.8
### Added
- Add `install_dependency` node option [#20](https://github.com/k0kubun/itamae-plugin-recipe-rbenv/pull/20)
  *Thanks to @sue445*

## v0.6.7
### Fixed
- Fix package installation error on Debian [#19](https://github.com/k0kubun/itamae-plugin-recipe-rbenv/pull/19)
  *Thanks to @hanachin*

## v0.6.6
### Fixed
- Support installing 2.1.10 when 2.1.1 is installed as well [#17](https://github.com/k0kubun/itamae-plugin-recipe-rbenv/pull/17)
  *Thanks to @chiastolite*

## v0.6.5
### Fixed
- Support Ubuntu 18.04 [#18](https://github.com/k0kubun/itamae-plugin-recipe-rbenv/pull/18)
  *Thanks to @swanmatch*

## v0.6.4
### Added
- Suport Amazon Linux [#16](https://github.com/k0kubun/itamae-plugin-recipe-rbenv/pull/16)
  *Thanks to @kawakubox*

## v0.6.3
### Fixed
- Fix the group of default-gems file [#14](https://github.com/k0kubun/itamae-plugin-recipe-rbenv/pull/14).
  *Thanks to @iyuuya*

## v0.6.2
### Fixed

- Fix the owner of user repository to rbenv:user [#13](https://github.com/k0kubun/itamae-plugin-recipe-rbenv/pull/13).
  *Thanks to @iyuuya*

## v0.6.1
### Added

- Support package download caching
  - Set `node[:rbenv][:cache]` true
  - See: https://github.com/rbenv/ruby-build#package-download-caching

## v0.6.0
### Changed

- Drop support for rbenv-gem-rehash since it's merged to rbenv core
  - See: https://github.com/rbenv/rbenv/pull/638
- Don't check base-devel group installation for Arch Linux
  - It's unnecessary for specinfra >= 2.50.2
  - See: https://github.com/mizzy/specinfra/pull/517

## v0.5.0
### Added

- Add support for Arch Linux, Linux Mint, OpenSUSE

### Fixed

- Add missing bzip2 dependency for RedHat, CentOS, Fedora

## v0.4.1
### Added

- rbenv's revision can be configurable with `node[:rbenv][:revision]`
  - To utilize optimization by https://github.com/itamae-kitchen/itamae/pull/182

## v0.4.0
### Added

- Support user-local rbenv installation by `rbenv::user`
- `rbenv_plugin` resource

### Changed

- Unused rbenv plugins are not `git clone`ed by default except ruby-build
  - You should explicitly set `install: true` if you want

## v0.3.4
### Added

- Add osx support for dependency installation.
  *Thanks to @dex1t*

## v0.3.3
### Added

- Add support for non-git scheme.
  *Thanks to @katsyoshi*

## v0.3.2
### Added

- Add support for rbenv-gem-rehash.
  *Thanks to @dex1t*

## v0.3.1
### Added

- Add git package installation.
  *Thanks to @takai*

## v0.3.0
### Changed

- Change platform finder from `os[:family]` to `node[:platform]`.
  *Thanks to @muratayusuke*
- Require itamae ~> v1.2

## v0.2.2
### Added

- Widen supported platforms for dependency installation.
  *Thanks to @sue445*

## v0.2.1
### Added

- Allow installing gems by default using rbenv-default-gems.
  *Thanks to @sue445*

## v0.2.0
### Added

- Allow specifying `RBENV_ROOT` by `node[:rbenv][:rbenv_root]`.
  *Thanks to @sue445*

## v0.1.2
### Fixed

- Add libffi-devel package depended by Ruby 2.2.
  *Thanks to @kwappa*

## v0.1.1
### Fixed

- Prevent error when `node[:'ruby-build']` is nil

## v0.1.0
### Added

- Support updating ruby-build's revision by `node[:'ruby-build'][:revision]`

## v0.0.1
### Added

- Clone rbenv and ruby-build by default
- Support ruby-build
  - installing multiple rubies using ruby-build
  - `rbenv global`


lib/itamae/plugin/recipe/rbenv/install.rb
# This recipe requires `rbenv_root` is defined.

if node[:rbenv][:install_development_dependency]
  include_recipe 'rbenv::development_dependency'
end
if node[:rbenv][:install_dependency]
  include_recipe 'rbenv::dependency'
end

scheme     = node[:rbenv][:scheme]
rbenv_root = node[:rbenv][:rbenv_root]

git rbenv_root do
  repository "#{scheme}://github.com/rbenv/rbenv.git"
  revision node[:rbenv][:revision] if node[:rbenv][:revision]
  user node[:rbenv][:user] if node[:rbenv][:user]
end

directory File.join(rbenv_root, 'plugins') do
  user node[:rbenv][:user] if node[:rbenv][:user]
end
if node[:rbenv][:cache]
  directory File.join(rbenv_root, 'cache') do
    user node[:rbenv][:user] if node[:rbenv][:user]
  end
end

define :rbenv_plugin, group: 'rbenv', revision: nil, install: nil do
  name = params[:name]
  group = params[:group]
  rev = params[:revision]
  if params[:install] == false
    install = false
  else
    install = true
  end

  if install || rev
    git "#{rbenv_root}/plugins/#{name}" do
      repository "#{scheme}://github.com/#{group}/#{name}.git"
      revision rev if rev
      user node[:rbenv][:user] if node[:rbenv][:user]
    end
  end
end

if node[:'rbenv-default-gems'] && node[:'rbenv-default-gems'][:'default-gems']
  rbenv_plugin 'rbenv-default-gems' do
    revision node[:'rbenv-default-gems'][:revision]
    install node[:'rbenv-default-gems'][:install]
  end

  node[:'rbenv-default-gems'][:install] = true
  file "#{rbenv_root}/default-gems" do
    content node[:'rbenv-default-gems'][:'default-gems'].join("\n") + "\n"
    mode    '664'
    if node[:rbenv][:user]
      owner node[:rbenv][:user]
      group node[:rbenv][:group] || node[:rbenv][:user]
    end
  end
end

rbenv_plugin 'ruby-build' do
  revision node[:'ruby-build'][:revision]
  install node[:'ruby-build'][:install]
end

rbenv_init = <<-EOS
  export RBENV_ROOT=#{rbenv_root}
  export PATH="#{rbenv_root}/bin:${PATH}"
  eval "$(rbenv init --no-rehash -)"
EOS

build_envs = node[:'ruby-build'][:build_envs].map do |key, value|
  %Q[export #{key}="#{value}"\n]
end.join

node[:rbenv][:versions].each do |version|
  execute "rbenv install #{version}" do
    command "#{rbenv_init} #{build_envs} rbenv install #{version}"
    not_if  "#{rbenv_init} rbenv versions --bare | grep -x #{version}"
    user node[:rbenv][:user] if node[:rbenv][:user]
  end
end

if node[:rbenv][:global]
  node[:rbenv][:global].tap do |version|
    execute "rbenv global #{version}" do
      command "#{rbenv_init} rbenv global #{version}"
      not_if  "#{rbenv_init} rbenv global | grep -x #{version}"
      user node[:rbenv][:user] if node[:rbenv][:user]
    end
  end
end

node[:rbenv][:plugins].each do |name, options|
  if name.include?('/')
    owner, repo = name.split('/', 2)
  else
    owner, repo = 'rbenv', name
  end

  rbenv_plugin repo do
    group owner
    revision options[:revision]
  end
end


lib/itamae/plugin/recipe/rbenv/dependency.rb
# Dependencies to install x.y.z (see also: development_dependency.rb)
# https://github.com/rbenv/ruby-build/wiki#suggested-build-environment
case node[:platform]
when 'debian', 'ubuntu', 'mint', 'raspbian'
  package 'build-essential'
  package 'libffi-dev'
  package 'libgdbm-dev'
  if node[:platform] == 'ubuntu' && node[:platform_version] >= '20.04'
      package 'libgdbm6'
      package 'libreadline-dev'
  elsif node[:platform] == 'ubuntu' && node[:platform_version] >= '18.04'
      package 'libgdbm5'
      package 'libreadline-dev'
  elsif node[:platform] == 'debian' && (node[:platform_version].to_f >= 10.0 || node[:platform_version].include?("testing"))
    package 'libgdbm6'
    package 'libreadline-dev'
  elsif node[:platform] == 'raspbian'
    package 'libgdbm6'
    package 'libreadline-dev'
  else
    package 'libgdbm3'
    package 'libreadline6-dev'
  end
  package 'libgmp-dev'
  package 'libncurses5-dev'
  package 'libssl-dev'
  package 'libyaml-dev'
  package 'rustc' # for yjit
  package 'zlib1g-dev'
when 'redhat', 'fedora', 'amazon' # redhat includes CentOS
  package 'bzip2'
  package 'gcc'
  package 'make'
  package 'gdbm-devel'
  package 'gmp-devel'
  package 'libffi-devel'
  package 'libyaml-devel'
  # rust package only provide after rhel 8
  if node[:platform] == 'redhat' && node[:platform_version].to_f > 8.0
    package 'rust' # for yjit
  end
  package 'ncurses-devel'
  package 'openssl-devel'
  package 'readline-devel'
  package 'zlib-devel'
when 'osx', 'darwin'
  package 'libffi'
  package 'libyaml'
  package 'openssl'
when 'arch'
  package 'base-devel'
  package 'gmp'
  package 'libffi'
  package 'libyaml'
  package 'openssl'
  package 'rust' # for yjit
  package 'zlib'
when 'opensuse'
  package 'bzip2'
  package 'gcc'
  package 'make'
  package 'gdbm-devel'
  package 'gmp-devel'
  package 'libffi-devel'
  package 'libyaml-devel'
  package 'ncurses-devel'
  package 'libopenssl-devel'
  package 'readline-devel'
  package 'rust' # for yjit
  package 'zlib-devel'
when 'gentoo'
  package 'bzip2'
  package 'gcc'
  package 'make'
  package 'gdbm'
  package 'libffi'
  package 'libyaml'
  package 'ncurses'
  package 'openssl'
  package 'readline'
  package 'zlib'
when 'freebsd'
  package 'lang/gcc'
  package 'databases/gdbm'
  package 'devel/gmake'
  package 'devel/gmp'
  package 'devel/libffi'
  package 'textproc/libyaml'
  package 'devel/ncurses'
  package 'security/openssl'
  package 'lang/rust' # for yjit
  package 'devel/readline'
when 'openbsd'
  package 'bzip2'
  package 'gdbm'
  package 'libffi'
  package 'libyaml'
  package 'openssl'
  package 'readline'
else
  raise NotImplementedError, "Platform '#{node[:platform]}' is not supported by rbenv::dependency yet"
end

package 'git'
